### ISSUE TITLE
How exactly does CallMolecularConsensusReads work?

### ISSUE THREAD

**Issue by noranekonobokkusu** [fulcrumgenomics/fgbio#898]
Hello,

I am trying to understand what CallMolecularConsensusReads is doing with my bam files. I have added the MI tag to my file using the `GroupReadsByUmi` command. Now I am trying to run `CallMolecularConsensusReads` on this file using different tags for the `-t` option. I decided to look at one particular UMI (UB:Z:TTTAAATTTTCC in the original file); it is associated with several batches of reads with different start coordinates. I am looking what `CallMolecularConsensusReads` produces for this barcode running it in three different ways:
1. with `-t MI` (the intended use, I believe) 
2. with `-t UB` (original barcodes, irregardless of start position)
3. with `-t UB_mod`, where UB_mod is made of the original UB field with the addition of start position (column 4 of the bam file, so the resulting UB field looks like UB:Z:TTTAAATTTTCC13950); I was trying this option hoping it will work identically to the generated MIs while preserving UMI sequences.

In contrast to my expectations, the statistics outputted for runs 1 and 2 look quite similar and are different from 3:

`-> java -jar fgbio.jar CallMolecularConsensusReads -t MI -i test.bam -o temp1.bam -M 3`
`[2023/01/18 03:40:22 | FgBioMain | Info] Executing CallMolecularConsensusReads from fgbio version 1.1.0 as safina@uger-c002.broadinstitute.org on JRE 11.0.16.1+1 with snappy, IntelInflater, and IntelDeflater`
`[2023/01/18 03:40:25 | CallMolecularConsensusReads | Info] Total Raw Reads Considered: 21,298.`
`[2023/01/18 03:40:25 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Orphan Consensus Created: 0 (0.0000).`
`[2023/01/18 03:40:25 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Mismatching Cigars: 103 (0.0048).`
`[2023/01/18 03:40:25 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Insufficient Support: 894 (0.0420).`
`[2023/01/18 03:40:25 | CallMolecularConsensusReads | Info] Consensus reads emitted: 1,219.`
`[2023/01/18 03:40:25 | FgBioMain | Info] CallMolecularConsensusReads completed. Elapsed time: 0.11 minutes.`

`-> java -jar fgbio.jar CallMolecularConsensusReads -t UB -i test.bam -o temp2.bam -M 3`
`[2023/01/18 03:40:46 | FgBioMain | Info] Executing CallMolecularConsensusReads from fgbio version 1.1.0 as safina@uger-c002.broadinstitute.org on JRE 11.0.16.1+1 with snappy, IntelInflater, and IntelDeflater`
`[2023/01/18 03:40:49 | CallMolecularConsensusReads | Info] Total Raw Reads Considered: 21,298.`
`[2023/01/18 03:40:49 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Orphan Consensus Created: 0 (0.0000).`
`[2023/01/18 03:40:49 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Mismatching Cigars: 103 (0.0048).`
`[2023/01/18 03:40:49 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Insufficient Support: 891 (0.0418).`
`[2023/01/18 03:40:49 | CallMolecularConsensusReads | Info] Consensus reads emitted: 1,215.`
`[2023/01/18 03:40:49 | FgBioMain | Info] CallMolecularConsensusReads completed. Elapsed time: 0.11 minutes.`

`-> java -jar fgbio.jar CallMolecularConsensusReads -t UB -i test2.bam -o temp3.bam -M 3`
`[2023/01/18 03:43:24 | FgBioMain | Info] Executing CallMolecularConsensusReads from fgbio version 1.1.0 as safina@uger-c002.broadinstitute.org on JRE 11.0.16.1+1 with snappy, IntelInflater, and IntelDeflater`
`[2023/01/18 03:43:26 | CallMolecularConsensusReads | Info] Total Raw Reads Considered: 24,189.`
`[2023/01/18 03:43:27 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Orphan Consensus Created: 0 (0.0000).`
`[2023/01/18 03:43:27 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Mismatching Cigars: 8 (0.0003).`
`[2023/01/18 03:43:27 | CallMolecularConsensusReads | Info] Raw Reads Filtered Due to Insufficient Support: 23,221 (0.9600).`
`[2023/01/18 03:43:27 | CallMolecularConsensusReads | Info] Consensus reads emitted: 257.`
`[2023/01/18 03:43:27 | FgBioMain | Info] CallMolecularConsensusReads completed. Elapsed time: 0.10 minutes.`

In the third run, the majority of reads were just discarded even though they were kept in the first run.
If I look at results for UB:Z:TTTAAATTTTCC UMI in the second run, I will find three consensus reads, although the UB field is the same for several reads having different start positions, and only one of the three consensus reads is messed up. Why is that?

I would highly appreciate it if you could explain what is happening, I feel rather lost at this point. I am attaching [test.bam.zip](https://github.com/fulcrumgenomics/fgbio/files/10441962/test.bam.zip) and [test2.bam.zip](https://github.com/fulcrumgenomics/fgbio/files/10441964/test2.bam.zip)

Thank you!

**Comment by nh13** [comment]
`GroupReadsByUmi` does not use the alignment start position of the read to group reads and assign them to the same source molecule.  From the [documentation](http://fulcrumgenomics.github.io/fgbio/tools/latest/GroupReadsByUmi.html):
>Reads are grouped by template, and then templates are sorted by the 5’ mapping positions of the reads from the template, used from earliest mapping position to latest. Reads that have the same end positions are then sub-grouped by UMI sequence.

The mapping position is the _unclipped_ position, where the alignment start (and end) are adjusted based on any leading (or trailing) unaligned (e.g. soft-clipped) bases.  Hence why you may see reads that have the same `MI` tag but not the same alignments in the SAM/BAM file.

Hopefully this helps.

**Comment by noranekonobokkusu** [comment]
Hi Nils, 

thank you! I'am afraid I overloaded my question with details, let me try again.
I am mainly wondering about how `CallMolecularConsensusReads` works with the `-t` tag I am providing. My understanding was that it should produce one consensus per unique tag (which I provide as the `-t` option). Still, if I run it as follows
`fgbio CallMolecularConsensusReads -t UB -I test.bam -o result.bam`
where UB is the field that contains just the UMI sequence UB:Z:TTTAAATTTTCC, I am getting three consensuses that have the same UB tag.

**Comment by noranekonobokkusu** [comment]
Thank you very much for clarifying this Nils, I understand what happened in my case now!
